lib/rubygems/gemcutter_utilities.rb

```diff
@@ -119,6 +119,40 @@ def sign_in sign_in_host = nil
     end
   end
 
+  ##
+  # Fetch user's multifactor authentication settings and return if extra digit
+  # code is needed.
+
+  def need_mfa?
+    return false if options[:suppress_mfa]
+    unless instance_variable_defined? :@mfa_level
+      response = rubygems_api_request(:get, 'api/v1/multifactor_auth') do |request|
```

Comment:
Won't this be adding additional api call to all requests? For default case, we should assume that mfa on api is disabled. mfa is an opt-in feature.
From server, we can return 403 or something more appropriate for cases where otp was required but not provided. Only then, client will prompt for otp and repeat the request.

Reply:
Here `options[:supress_mfa]` is just an internal option for keep compatibility with older test cases, since many of them relies on `last_request`. And fail message the client received is `plain/text`, should I just match it literally?

Reply:
> for keep compatibility with older test cases, since many of them relies on last_request

Can you please explain more about this limitation? Are you sure we can't do away with `options[:suppress_mfa]`?

>  should I just match it literally?

Yeah, checking for response text would be good enough unless we come up with something more elegant.

Reply:
If we want to remove `options[:suppress_mfa]`, I have tried that create a new module dynamically, setting `need_mfa?` always return false, and prepend it to `Gem::Commands::PushCommand` in test cases which will complain if mfa related requests are added. This method should be not so intrusive, like [here](https://paste.ofcode.org/3aBhsqT54rrcmeZUfjsBdTh).